packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserver.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "PerformanceObserver.h"
+#include "PerformanceObserverRegistry.h"
+#include "PerformanceEntryReporter.h"
+
+namespace facebook::react {
+
+PerformanceObserver::~PerformanceObserver() {
+  if (auto registry = registry_.lock()) {
+    registry->removeObserver(*this);
+  }
+}
+
+void PerformanceObserver::pushEntry(const PerformanceEntry& entry) {
+  buffer_.add(entry);
+}
+
+std::vector<PerformanceEntry> PerformanceObserver::takeRecords() {
+  return buffer_.consume();
+}
+
+bool PerformanceObserver::isObserving(PerformanceEntryType type) const {
+  return observedTypes_.contains(type);
+}
+
+void PerformanceObserver::observe(PerformanceEntryType type, bool buffered) {
+  // we assume that `type` was checked on JS side and is correct
+  observedTypes_.clear();
+  observedTypes_.insert(type);
+
+  requiresDroppedEntries_ = true;
```

Comment:
I was going to suggest a better name for this (using "buffered" instead of "dropped", because the actual dropped ones are irrecoverable) but then I realized we don't need this variable at all.

We should be collecting the buffered entries immediately here as per the spec. This works on Web:

```
var observer = new PerformanceObserver(() => {});
observer.observe({type: 'mark', buffered: true});
observer.takeRecords(); // returns all buffered entries
```

Reply:
The goal is was match spec naming, so it is easier to maintain in the future.